runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/AbstractWindmillStream.java

```diff
@@ -255,6 +266,28 @@ public final Instant startTime() {
     return new Instant(startTimeMs.get());
   }
 
+  @Override
+  public String backendWorkerToken() {
+    return backendWorkerToken;
+  }
+
+  @Override
+  public void shutdown() {
+    if (isShutdown.compareAndSet(false, true)) {
+      halfClose();
```

Comment:
it's not clear to me we want to halfclose here, it seems it may be better to do requestObserver.onError(...)

also if we've already called halfclose it may be invalid to call it again.

Reply:
done